### ISSUE TITLE
[zuo] Chez Scheme test failures

### ISSUE THREAD

**Issue by LiberalArtist** [racket/racket#4223]
Following up on https://github.com/racket/racket/pull/4203#issuecomment-1112575524 and [my Discourse asside](https://racket.discourse.group/t/chez-for-architectures-without-native-backends/950/5?u=liberalartist), I've tried an out-of-source build of https://github.com/racket/racket/commit/10a34f3007eae594b363fcfe11c6b2e02c51eeee with `#:parallel-tests? #f` at the Guix level (so, no `-j` for `make`), and the Chez Scheme test suite failed with errors that, at least so far, look pretty much the same as before. For the record, the Guix package definition I used is in https://gitlab.com/philip1/guix-patches/-/commit/c7c94e29a65f7a83ccf702cd965bafb54e4f2c88 (the `zuo` branch extends the `racket-8.5` branch in that repository), and I've put the full log, environment variables file, and build directory (33.4 MiB Zstd-compressed) at <https://philipmcgrath.com/tmp/zuo-chez-scheme-for-racket-c7c94e29a65f7a83ccf702cd965bafb54e4f2c88.tar.zst>.

I will try to look at `mat.ss` as you suggested in https://github.com/racket/racket/pull/4203#issuecomment-1112575524 next and see if anything strikes me as a possible cause of Guix weirdness, though I've never really looked at the Chez test suite before beyond running `make test`.

**Comment by samth** [comment]
If you run just the individual mats that fail, do they behave the same?

**Comment by LiberalArtist** [comment]
One of the early failures:
```
< profile.mo:Bug in mat compile-profile clause 134 at line 612, char 3 of /tmp/guix-build-chez-scheme-for-racket-9.5.7.7.drv-0/source/racket/src/ChezScheme/mats/profile.ms
```
points to this code:
https://github.com/racket/racket/blob/1eec5bdb09ab812a7e6129a254640b53fba9df3e/racket/src/ChezScheme/mats/profile.ms#L612-L616

I still don't understand what's going on, but I see that `testfile-cp3.ss` is created, but `testfile-cp3.so` seems not to be.

I'm also a bit suspicious of comparing the result of `source-file-descriptor-path` to `"testfile-ca3.ss"` with `equal?`. I haven't figured out a good way to experiment yet, but CSUG [§ 11.1 "Annotations"](https://cisco.github.io/ChezScheme/csug9.5/syntax.html#./syntax:h11) says that "the path [of a source-file descriptor] might or might not be an absolute path depending on how the file's path was specified when the source-file descriptor was created," and the [examples for `with-source-path`](https://cisco.github.io/ChezScheme/csug9.5/system.html#./system:s103) seem to suggest that paths may be sensitive to the working directory and the way the paths set as `source-directories` are written.

Re @samth:
> If you run just the individual mats that fail, do they behave the same?

I'm still looking at this. It seems like all of the configurations (in the sense of `IMPLEMENTATION.md`'s ["Running Tests in One Configuration (with expected-error checking)"](https://github.com/racket/racket/blob/master/racket/src/ChezScheme/IMPLEMENTATION.md#running-tests-in-one-configuration-with-expected-error-checking)) have errors. If the same sets tests (in the sense of ["Running One Set of Tests (no expected-error checking)"](https://github.com/racket/racket/blob/master/racket/src/ChezScheme/IMPLEMENTATION.md#running-one-set-of-tests-no-expected-error-checking)—I think this is what you meant) are failing in all configurations, that does seem worth trying, but I think I would need to figure out how to do so and evaluate the results usefully with Guix. I think I had not encountered these failures outside of the Guix package build environment, but I'm not completely certain I've tried a comparable setup.

**Comment by mflatt** [comment]
If the failures involve `separate-eval`, then I wonder whether it could be a problem communicating the current directory in the original Chez Scheme process to the new and separate Chez Scheme process. Chez Scheme sets the directory with `chdir`, so that should get propagated to a subprocess, but maybe something about starting a Chez Scheme instance adjusts the directory.

> I'm also a bit suspicious of comparing the result of `source-file-descriptor-path` to "testfile-ca3.ss" with `equal?`

The tests are set up to make that predictable, but it's a little fragile and was a big source of trouble when adjusting the build system and tests to work without cloning sources to a build directory. My guess is that this isn't the main problem for the failures you're seeing, but I could be wrong.

To run just `profile.ms`, for example: in the `ta6le/mats` build directory, use `zuo . profile.mo`. If won't check for expected errors, but it will look for `Bug` output.

**Comment by LiberalArtist** [comment]
I've tried replacing the phase that runs `make test` with the equivalent of `cd ta6le/mats; zuo . profile.mo`, which produced this error:

```
phase `build' succeeded after 33.0 seconds
starting phase `check'
running ../bin/ta6le/scheme to build compile
compiling ../../../ChezScheme/mats/mat.ss with output to mat.so
running ../bin/ta6le/scheme to build profile.mo with output to profile.out
Bug in mat compile-profile clause 134 at line 612, char 3 of ../../../ChezScheme/mats/profile.ms
failed
 in build-one
 in loop
 in module->hash
error: in phase 'check': uncaught exception:
%exception #<&invoke-error program: "zuo" arguments: ("." "profile.mo") exit-status: 1 term-signal: #f stop-signal: #f> 
phase `check' failed after 0.3 seconds
command "zuo" "." "profile.mo" failed with status 1
note: keeping build directory `/tmp/guix-build-chez-scheme-for-racket-9.5.7.7.drv-1'
builder for `/gnu/store/szq7lys3pxc0si7xxy5h0zxpipdymb1m-chez-scheme-for-racket-9.5.7.7.drv' failed with exit code 1
build of /gnu/store/szq7lys3pxc0si7xxy5h0zxpipdymb1m-chez-scheme-for-racket-9.5.7.7.drv failed
View build log at '/var/log/guix/drvs/sz/q7lys3pxc0si7xxy5h0zxpipdymb1m-chez-scheme-for-racket-9.5.7.7.drv.gz'.
cannot build derivation `/gnu/store/xmwpm8q54ry1sli03s3w0daa0snqisw8-racket-vm-cs-8.5.0.2.drv': 1 dependencies couldn't be built
cannot build derivation `/gnu/store/dp6c1waxw6g6v71z6ai7wp57lw8aps1x-racket-8.5.0.2.drv': 1 dependencies couldn't be built
guix build: error: build of `/gnu/store/dp6c1waxw6g6v71z6ai7wp57lw8aps1x-racket-8.5.0.2.drv' failed
```

Here are [`profile.out.log`](https://github.com/racket/racket/files/8691036/profile.out.log) and [`profile.mo.log`](https://github.com/racket/racket/files/8691037/profile.mo.log) (with the `.log` extension added to placate GitHub).

This all looks pretty similar to previous results, so it seems parallel tests were not the problem (or at least not the only problem).

This Guix definition for this build is at https://gitlab.com/philip1/guix-patches/-/commit/99f0b2f55874dcacaaa8768c5732b29f277a88fb: it uses 1eec5bdb09ab812a7e6129a254640b53fba9df3e.

**Comment by LiberalArtist** [comment]
This got a bit further! Building https://github.com/racket/racket/commit/9f2b802acdcf9187b4c46fb5a9721a7f716952ca with the patch for `s_process` from https://github.com/LiberalArtist/racket/commit/8029009e40fbe1008d08e20aae5889a71b7c7d53 (all in the Guix definitions at https://gitlab.com/philip1/guix-patches/-/commit/e9c46f5a8f6602d9ac0a6eb0c7d72a23d3d2c8f0) produced some new errors:
```
phase `build' succeeded after 33.0 seconds
starting phase `check'
running ../bin/ta6le/scheme to build compile
compiling ../../../ChezScheme/mats/mat.ss with output to mat.so
running ../bin/ta6le/scheme to build profile.mo with output to profile.out
Error in mat compile-profile clause 133: "separate-compile: Exception: library (testfile-ca3) not found " at line 592, char 3 of ../../../ChezScheme/mats/profile.ms
Error in mat compile-profile clause 134: "separate-eval: Exception in load-program: failed for testfile-cp3.so: no such file or directory Exception in cdr: #f is not a pair " at line 612, char 3 of ../../../ChezScheme/mats/profile.ms
failed
 in build-one
 in loop
 in module->hash
error: in phase 'check': uncaught exception:
%exception #<&invoke-error program: "zuo" arguments: ("." "profile.mo") exit-status: 1 term-signal: #f stop-signal: #f> 
phase `check' failed after 0.4 seconds
command "zuo" "." "profile.mo" failed with status 1
note: keeping build directory `/tmp/guix-build-chez-scheme-for-racket-9.5.7.7.drv-0'
builder for `/gnu/store/czgm36lcp7x5xs6qlchzm15lky8fvs8b-chez-scheme-for-racket-9.5.7.7.drv' failed with exit code 1
build of /gnu/store/czgm36lcp7x5xs6qlchzm15lky8fvs8b-chez-scheme-for-racket-9.5.7.7.drv failed
View build log at '/var/log/guix/drvs/cz/gm36lcp7x5xs6qlchzm15lky8fvs8b-chez-scheme-for-racket-9.5.7.7.drv.gz'.
guix build: error: build of `/gnu/store/czgm36lcp7x5xs6qlchzm15lky8fvs8b-chez-scheme-for-racket-9.5.7.7.drv' failed
```
Here are [gm36lcp7x5xs6qlchzm15lky8fvs8b-chez-scheme-for-racket-9.5.7.7.drv.gz](https://github.com/racket/racket/files/8734427/gm36lcp7x5xs6qlchzm15lky8fvs8b-chez-scheme-for-racket-9.5.7.7.drv.gz) and (again with `.log` to placate GitHub) [profile.out.log](https://github.com/racket/racket/files/8734445/profile.out.log) and [profile.mo.log](https://github.com/racket/racket/files/8734446/profile.mo.log).

Notably, `testfile-ca3.so` and `testfile-ca3.ss` both exist in the `mats` build directory:
```
philip@bastet:/tmp/guix-build-chez-scheme-for-racket-9.5.7.7.drv-0/source/racket/src/build/ta6le/mats$ tree
.
├── main.zuo
├── mat.so
├── profile.html
├── profile.mo
├── profile.mo.log
├── profile.out
├── profile.out.log
├── testdir
│   ├── profile.html
│   └── testfile.ss.html
├── testfile1.pd
├── testfile2.pd
├── testfile3.pd
├── testfile-block.so
├── testfile-block.ss
├── testfile-ca3.so
├── testfile-ca3.ss
├── testfile-coverage1a.covin
├── testfile-coverage1a.so
├── testfile-coverage1a.ss
├── testfile-coverage1b.covin
├── testfile-coverage1b.so
├── testfile-coverage1b.ss
├── testfile-coverage1c.covin
├── testfile-coverage1c.so
├── testfile-coverage1c.ss
├── testfile-coverage1d.covin
├── testfile-coverage1d.so
├── testfile-coverage1d.ss
├── testfile.covout
├── testfile-cp1.ss
├── testfile-cp2.so
├── testfile-cp2.ss
├── testfile-cp3.ss
├── testfile-cp4.so
├── testfile-cp4.ss
├── testfile.so
├── testfile.ss
├── testfile.ss.html
├── testfile-triv.so
├── testfile-triv.ss
└── _zuo.db

1 directory, 41 files
```

This seems consistent with @mflatt's comment:

> If the failures involve `separate-eval`, then I wonder whether it could be a problem communicating the current directory in the original Chez Scheme process to the new and separate Chez Scheme process. Chez Scheme sets the directory with `chdir`, so that should get propagated to a subprocess, but maybe something about starting a Chez Scheme instance adjusts the directory.

**Comment by LiberalArtist** [comment]
I think I've tracked down the issue, but I'm not sure why it hadn't been a problem before, nor whether it's a scenario the Zuo scripts ought to manage or not.

I noticed that Guix was generating the following in its `sh` script to initialize environment variables in the build container:
```sh
export CHEZSCHEMELIBDIRS="/gnu/store/z8i94qpg75zf6gxkp60pib815l080a9j-chez-nanopass-1.9.2/lib/chez-scheme"
```
Because the value of `CHEZSCHEMELIBDIRS` did not contain `.` and did not end with `:`, running `scheme` without the `--libdirs` flag would not include the current directory in the initial value of `(library-directories)`.

Adjusting the Guix packaging to unset `CHEZSCHEMELIBDIRS` was enough to fix at least `zuo . profile.mo`: https://gitlab.com/philip1/guix-patches/-/commit/530242e1928c07a32a9d7207e975cd4cbca47800 I'm currently running another build with the full test suite.

For Guix, it is certainly good enough to unset `CHEZSCHEMELIBDIRS`. (It is set automatically as a side-effect of the way we have arranged to share the Nanopass source code among multiple packages.) I expect it would be possible to make things work with such a value for `CHEZSCHEMELIBDIRS` by changing the Zuo scripts and `separate-eval`/`separate-compile`, but I'm not sure if that would be worth-while, especially since it seems like changes would probably be needed in multiple places. In particular, I don't know if Guix is doing the right thing, or at least a reasonable thing, by not including `.` in its synthesized definitions for `CHEZSCHEMELIBDIRS`, either implicitly or explicitly. I know Guix package definitions for other Chez libraries have to add `.`, which makes me especially suspicious. But personally I only use Chez in relation to Racket or when comparing Scheme implementations, not as a normal programming environment.

If Chez users more generally are at least reasonably likely to have `CHEZSCHEMELIBDIRS` set without an entry for `.`, then perhaps it's worth making this work, but if this is a Guix idiosyncrasy, then we should deal with it in Guix.

**Comment by mflatt** [comment]
Thanks for tracking this down! The old `mats/Mf-base` set `CHEZSCHEMELIBDIRS` to `.`, so that's probably why things worked before. Along those lines, I think it does make sense for `mats/build.zuo` to set `CHEZSCHEMELIBDIRS` (or remove a definition from the environment) before running. I'll make that change.

**Comment by LiberalArtist** [comment]
The build with all tests enabled got to a new error:
```
gcc -DGUIX_RKTIO_PATCH_BIN_SH=/gnu/store/chfwin3a4qp1znnpsjbmydr2jbzk0d6y-bash-minimal-5.1.8/bin/sh -Ita6le/boot/ta6le -Ita6le/mats -m64 -msse2 -O2 -D_REENTRANT -pthread -fPIC -shared -o ta6le/mats/foreign1.so ../ChezScheme/mats/foreign1.c ../ChezScheme/mats/foreign2.c ../ChezScheme/mats/foreign3.c ../ChezScheme/mats/foreign4.c -rdynamic -lm -ldl -lncurses -lrt -lpthread
gcc -DGUIX_RKTIO_PATCH_BIN_SH=/gnu/store/chfwin3a4qp1znnpsjbmydr2jbzk0d6y-bash-minimal-5.1.8/bin/sh -Ita6le/boot/ta6le -Ita6le/mats -m64 -msse2 -O2 -D_REENTRANT -pthread -o ta6le/mats/cat_flush ../ChezScheme/mats/cat_flush.c -rdynamic -lm -ldl -lncurses -lrt -lpthread
running ta6le/bin/ta6le/scheme to build all with output to ta6le/mats/output-compile-0-f-f-f-simple/make.out
running ta6le/bin/ta6le/scheme to build all with output to ta6le/mats/output-compile-3-f-f-f-simple/make.out
patch ta6le/mats/experr-compile-0-f-t-f ../ChezScheme/mats/patch-compile-0-f-t-f
patching file ta6le/mats/experr-compile-0-f-t-f
running ta6le/bin/ta6le/scheme to build all with output to ta6le/mats/output-compile-0-f-t-f-cl3/make.out
patch ta6le/mats/experr-compile-3-f-t-f ../ChezScheme/mats/patch-compile-3-f-t-f
running ta6le/bin/ta6le/scheme to build all with output to ta6le/mats/output-compile-3-f-t-f-cl3/make.out
patch ta6le/mats/experr-compile-0-t-f-f ../ChezScheme/mats/patch-compile-0-t-f-f
patching file ta6le/mats/experr-compile-0-t-f-f
running ta6le/bin/ta6le/scheme to build all with output to ta6le/mats/output-compile-0-t-f-f-spi-rmg/make.out
patch ta6le/mats/experr-compile-3-t-f-f ../ChezScheme/mats/patch-compile-3-t-f-f
running ta6le/bin/ta6le/scheme to build all with output to ta6le/mats/output-compile-3-t-f-f-spi-rmg/make.out
patch ta6le/mats/experr-interpret-0-f-f-f ../ChezScheme/mats/patch-interpret-0-f-f-f
patching file ta6le/mats/experr-interpret-0-f-f-f
running ta6le/bin/ta6le/scheme to build all with output to ta6le/mats/output-interpret-0-f-f-f-cl6/make.out
patch ta6le/mats/experr-interpret-3-f-f-f ../ChezScheme/mats/patch-interpret-3-f-f-f
running ta6le/bin/ta6le/scheme to build all with output to ta6le/mats/output-interpret-3-f-f-f-cl6/make.out
patch ta6le/mats/experr-interpret-0-f-t-f ../ChezScheme/mats/patch-interpret-0-f-t-f
patching file ta6le/mats/experr-interpret-0-f-t-f
running ta6le/bin/ta6le/scheme to build all with output to ta6le/mats/output-interpret-0-f-t-f-rmg2/make.out
patch ta6le/mats/experr-interpret-3-f-t-f ../ChezScheme/mats/patch-interpret-3-f-t-f
running ta6le/bin/ta6le/scheme to build all with output to ta6le/mats/output-interpret-3-f-t-f-rmg2/make.out
running ta6le/bin/ta6le/scheme to build all with output to ta6le/mats/output-compile-0-f-f-f-hci-cl/make.out
running ta6le/bin/ta6le/scheme to build all with output to ta6le/mats/output-interpret-3-f-f-f-hci-rmg/make.out
-------- o=0 eoc=f simple --------
10317d10316
< ftype.mo:Error in mat ftype clause 71: "ftype-test: C compilation failed" at line 630, char 3 of /tmp/guix-build-chez-scheme-for-racket-9.5.7.7.drv-0/source/racket/src/ChezScheme/mats/ftype.ms
-------- o=3 eoc=f simple --------
27d26
< ftype.mo:Error in mat ftype clause 71: "ftype-test: C compilation failed" at line 630, char 3 of /tmp/guix-build-chez-scheme-for-racket-9.5.7.7.drv-0/source/racket/src/ChezScheme/mats/ftype.ms
-------- o=0 cp0=t cl=3 cl3 --------
10317d10316
< ftype.mo:Error in mat ftype clause 71: "ftype-test: C compilation failed" at line 630, char 3 of /tmp/guix-build-chez-scheme-for-racket-9.5.7.7.drv-0/source/racket/src/ChezScheme/mats/ftype.ms
-------- o=3 cp0=t cl=3 cl3 --------
27d26
< ftype.mo:Error in mat ftype clause 71: "ftype-test: C compilation failed" at line 630, char 3 of /tmp/guix-build-chez-scheme-for-racket-9.5.7.7.drv-0/source/racket/src/ChezScheme/mats/ftype.ms
-------- o=0 spi=t p=t rmg=2 spi-rmg --------
10317d10316
< ftype.mo:Error in mat ftype clause 71: "ftype-test: C compilation failed" at line 630, char 3 of /tmp/guix-build-chez-scheme-for-racket-9.5.7.7.drv-0/source/racket/src/ChezScheme/mats/ftype.ms
-------- o=3 spi=t p=t rmg=2 spi-rmg --------
27d26
< ftype.mo:Error in mat ftype clause 71: "ftype-test: C compilation failed" at line 630, char 3 of /tmp/guix-build-chez-scheme-for-racket-9.5.7.7.drv-0/source/racket/src/ChezScheme/mats/ftype.ms
-------- o=0 eval=interpret cl=6 cl6 --------
10317d10316
< ftype.mo:Error in mat ftype clause 71: "ftype-test: C compilation failed" at line 630, char 3 of /tmp/guix-build-chez-scheme-for-racket-9.5.7.7.drv-0/source/racket/src/ChezScheme/mats/ftype.ms
-------- o=3 eval=interpret cl=6 cl6 --------
27d26
< ftype.mo:Error in mat ftype clause 71: "ftype-test: C compilation failed" at line 630, char 3 of /tmp/guix-build-chez-scheme-for-racket-9.5.7.7.drv-0/source/racket/src/ChezScheme/mats/ftype.ms
-------- o=0 cp0=t eval=interpret rmg=2 rmg2 --------
10317d10316
< ftype.mo:Error in mat ftype clause 71: "ftype-test: C compilation failed" at line 630, char 3 of /tmp/guix-build-chez-scheme-for-racket-9.5.7.7.drv-0/source/racket/src/ChezScheme/mats/ftype.ms
-------- o=3 cp0=t eval=interpret rmg=2 rmg2 --------
27d26
< ftype.mo:Error in mat ftype clause 71: "ftype-test: C compilation failed" at line 630, char 3 of /tmp/guix-build-chez-scheme-for-racket-9.5.7.7.drv-0/source/racket/src/ChezScheme/mats/ftype.ms
-------- o=0 hci=101 eoc=f cl=9 hci-cl --------
10317d10316
< ftype.mo:Error in mat ftype clause 71: "ftype-test: C compilation failed" at line 630, char 3 of /tmp/guix-build-chez-scheme-for-racket-9.5.7.7.drv-0/source/racket/src/ChezScheme/mats/ftype.ms
-------- o=3 hci=101 eval=interpret rmg=2 hci-rmg --------
27d26
< ftype.mo:Error in mat ftype clause 71: "ftype-test: C compilation failed" at line 630, char 3 of /tmp/guix-build-chez-scheme-for-racket-9.5.7.7.drv-0/source/racket/src/ChezScheme/mats/ftype.ms
some output differs from expected
 in build-one
 in loop
 in module->hash
make: *** [Makefile:30: test] Error 1

Test suite failed, dumping logs.
error: in phase 'check': uncaught exception:
%exception #<&invoke-error program: "make" arguments: ("test" "-j" "16" "ZUO=/gnu/store/i7a07a1cpii349w13hl42idinavhzbm9-zuo-1-racket8.5.0.2-guix1/bin/zuo" "STEXLIB=/gnu/store/01q09dm32c097w3672wkpmfmwkxqx3s7-stex-1.2.2-3.9434b4a/lib/stex") exit-status: 2 term-signal: #f stop-signal: #f> 
phase `check' failed after 3093.0 seconds
command "make" "test" "-j" "16" "ZUO=/gnu/store/i7a07a1cpii349w13hl42idinavhzbm9-zuo-1-racket8.5.0.2-guix1/bin/zuo" "STEXLIB=/gnu/store/01q09dm32c097w3672wkpmfmwkxqx3s7-stex-1.2.2-3.9434b4a/lib/stex" failed with status 2
note: keeping build directory `/tmp/guix-build-chez-scheme-for-racket-9.5.7.7.drv-1'
builder for `/gnu/store/g3x37f63ghmh3nlw58s9gyi4wxagwvsc-chez-scheme-for-racket-9.5.7.7.drv' failed with exit code 1
build of /gnu/store/g3x37f63ghmh3nlw58s9gyi4wxagwvsc-chez-scheme-for-racket-9.5.7.7.drv failed
View build log at '/var/log/guix/drvs/g3/x37f63ghmh3nlw58s9gyi4wxagwvsc-chez-scheme-for-racket-9.5.7.7.drv.gz'.
guix build: error: build of `/gnu/store/g3x37f63ghmh3nlw58s9gyi4wxagwvsc-chez-scheme-for-racket-9.5.7.7.drv' failed
```
Here is the full build log: [x37f63ghmh3nlw58s9gyi4wxagwvsc-chez-scheme-for-racket-9.5.7.7.drv.gz](https://github.com/racket/racket/files/8766372/x37f63ghmh3nlw58s9gyi4wxagwvsc-chez-scheme-for-racket-9.5.7.7.drv.gz)

and here is files from `racket/src/build/ta6le/mats/summary`, with `.log` to placate GitHub: [summary.log](https://github.com/racket/racket/files/8766391/summary.log).

It looks like the error is coming from:
https://github.com/racket/racket/blob/9f2b802acdcf9187b4c46fb5a9721a7f716952ca/racket/src/ChezScheme/mats/ftype.ms#L549-L577

where I think the issue is that `cc` isn't on the `PATH` in the build environment:
```
philip@bastet:~/code/guix/racket$ ls -l /gnu/store/069aq2v993kpc41yabp5b6vm4wb9jkhg-gcc-10.3.0/bin
total 21422
-r-xr-xr-x 5 root root  1214912 Dec 31  1969 c++
-r-xr-xr-x 2 root root  1214912 Dec 31  1969 cpp
-r-xr-xr-x 5 root root  1214912 Dec 31  1969 g++
-r-xr-xr-x 4 root root  1214912 Dec 31  1969 gcc
-r-xr-xr-x 3 root root    35256 Dec 31  1969 gcc-ar
-r-xr-xr-x 3 root root    35256 Dec 31  1969 gcc-nm
-r-xr-xr-x 3 root root    35256 Dec 31  1969 gcc-ranlib
-r-xr-xr-x 2 root root   724672 Dec 31  1969 gcov
-r-xr-xr-x 2 root root   585304 Dec 31  1969 gcov-dump
-r-xr-xr-x 2 root root   605848 Dec 31  1969 gcov-tool
-r-xr-xr-x 2 root root 27434408 Dec 31  1969 lto-dump
-r-xr-xr-x 5 root root  1214912 Dec 31  1969 x86_64-unknown-linux-gnu-c++
-r-xr-xr-x 5 root root  1214912 Dec 31  1969 x86_64-unknown-linux-gnu-g++
-r-xr-xr-x 4 root root  1214912 Dec 31  1969 x86_64-unknown-linux-gnu-gcc
-r-xr-xr-x 4 root root  1214912 Dec 31  1969 x86_64-unknown-linux-gnu-gcc-10.3.0
-r-xr-xr-x 3 root root    35256 Dec 31  1969 x86_64-unknown-linux-gnu-gcc-ar
-r-xr-xr-x 3 root root    35256 Dec 31  1969 x86_64-unknown-linux-gnu-gcc-nm
-r-xr-xr-x 3 root root    35256 Dec 31  1969 x86_64-unknown-linux-gnu-gcc-ranlib
```

I can try a workaround from Guix, but it seems like it might be better in general to let that logic benefit from `./configure` and the build system's knowledge of how to compile C code.

The Guix definitions for this build are at: https://gitlab.com/philip1/guix-patches/-/commit/922e27ec3ff20bf2872bcfa57be9a290ab08a12b

**Comment by mflatt** [comment]
That should definitely change to use the configured compiler and flags, and I've pushed the change.

**Comment by LiberalArtist** [comment]
Adding `cc` as a symlink fixed the build of 3e17c5a all the way through Guix's `racket` package (i.e. `main-distribution`): https://gitlab.com/philip1/guix-patches/-/commit/b435128c82e399a036559106408c650bb81884ca 

I'll try building 012646c next.

**Comment by LiberalArtist** [comment]
And everything works with 012646c! https://gitlab.com/philip1/guix-patches/-/commit/271fc335d613c0d34fd02cbcd495ee3a4be21b40
